server/survey.go

```diff
@@ -30,80 +29,105 @@ const (
-func (p *Plugin) checkForNextSurvey(currentVersion semver.Version) bool {
+func (p *Plugin) checkForNextSurvey(now time.Time) bool {
 	// Add a random delay to mitigate against the fact that multiple instances of the plugin may be trying to start up across
 	// different servers and we have no real way to synchronize between them.
 	p.sleepUpTo(p.upgradeCheckMaxDelay)
 
+	// TODO replace this with a proper HA-friendly lock
 	p.surveyLock.Lock()
 	defer p.surveyLock.Unlock()
 
 	if !p.getConfiguration().EnableSurvey {
 		// Surveys are disabled, so return false without updating the stored version. If surveys are re-enabled, the
 		// plugin will then detect an upgrade (if one occurred) and schedule the next survey.
-		p.API.LogDebug("Not sending NPS survey because survey is disabled")
+		p.API.LogInfo("Not sending NPS survey because survey is disabled")
 		return false
 	}
 
-	lastUpgrade, _ := p.getLastServerUpgrade()
-
-	if !shouldScheduleSurvey(currentVersion, lastUpgrade) {
-		// No version change
-		p.API.LogDebug("No server version change detected. Not scheduling a new survey.")
+	var nextSurvey *surveyState
+	if err := p.KVGet(fmt.Sprintf(SURVEY_KEY, p.serverVersion), &nextSurvey); err != nil {
+		p.API.LogError("Failed to get survey state", "err", err)
 		return false
 	}
 
-	now := time.Now().UTC()
-	nextSurvey := now.Add(TIME_UNTIL_SURVEY)
+	if nextSurvey != nil {
+		p.API.LogInfo(fmt.Sprintf("Survey already scheduled for %s", nextSurvey.StartAt.Format("Jan 2, 2006")))
+		return false
+	}
 
-	if lastUpgrade == nil {
-		p.API.LogInfo(fmt.Sprintf("NPS plugin installed. Scheduling NPS survey for %s", nextSurvey.Format("Jan 2, 2006")))
-	} else {
-		p.API.LogInfo(fmt.Sprintf("Version change detected from %s to %s. Scheduling NPS survey for %s", lastUpgrade.Version, currentVersion, nextSurvey.Format("Jan 2, 2006")))
+	nextSurvey = &surveyState{
+		ServerVersion: p.serverVersion,
+		CreateAt:      now,
+		StartAt:       now.Add(TIME_UNTIL_SURVEY),
 	}
 
-	if shouldSendAdminNotices(now, lastUpgrade) {
-		p.sendAdminNotices(nextSurvey)
+	p.API.LogInfo(fmt.Sprintf("Scheduling next survey for %s", nextSurvey.StartAt.Format("Jan 2, 2006")))
+
+	if err := p.KVSet(fmt.Sprintf(SURVEY_KEY, p.serverVersion), nextSurvey); err != nil {
+		p.API.LogError("Failed to schedule next survey", "err", err)
+		return false
 	}
 
-	if err := p.storeServerUpgrade(&serverUpgrade{
-		Version:   currentVersion,
-		Timestamp: now,
-	}); err != nil {
-		p.API.LogError("Failed to store time of server upgrade. The next NPS survey may not occur.", "err", err)
+	if sent, err := p.sendAdminNotices(now, nextSurvey); err != nil {
+		p.API.LogError("Failed to send notification of next survey to admins", "err", err)
+		return false
+	} else if !sent {
+		p.API.LogInfo("Not sending notification of next survey to admins since they already received one recently")
+	} else {
+		p.API.LogInfo("Sent notification of next survey to admins")
```

Comment:
I wonder if these log messages should be part of `sendAdminNotices`? We're inferring a lot based on whether `sent` is true or not

Reply:
That's a good point. I think that would be a good idea, but I'd rather not move them right now because that'll require that I mock `LogInfo` in all of the tests in `sendAdminNotices`.